contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,504 @@
+
+### Unset fields:
+
+When a field is not specified it is automatically allowed. This makes it easy to rollout the feature for existing clusters, as it makes everything allowed in the cluster when no policy is created.
+
+### Inside the matches:
+
+At the policy level, the match field can express pretty much any structure you want to, but there's some things that should be considered:
+
+- the structure should match as much as possible what you try to take action on.
+- match elements are combined.
+
+example :
+
+```
+- match:
+  - keys: ["projectA-dedicated","projectB-dedicated"]
+    operators: ["Exists"]
+    effects: []
+```
+
+- This matches the structure of toleration
+- This match means the following : t
+   - The toleration `projectA-dedicated` with the operator `Exists`
+   - The toleration `projectB-dedicated` with the operator `Exists`
+
+### Policy composition and conflict handling
+
+Policies are composed by ANDing them, note that rules of policies from a lower priority are superseded by ones from a higher priority if there is a conflict.
+
+There is two kinds of conflict to handle in policies of the same priority level: Structural and semantical conflicts.
+
+
+Structural conflicts must handled at creation time as much as possible, on the other hand, semantical conflicts should be handled at runtime: detect that there's a conflict and emit an event stating that policies couldn't be satisfied due to a conflict.
+
+
+### How policies are computed
+
+Policies may have overlapping rules, to handle this policies are computed in the following order:
+
+- compute policies at `exception` priority.
+- compute policies at `cluster` priority.
+- compute policies at `user` priority.
+- compute policies at `default` priority.
+
+If a policy doesn't specify a priority the default priority applies.
+
+they should also obey to the following rules:
+
+- everything that is required is by definition allowed.
+
+
+# Detailed Design
+
+
+## SchedulingPolicy
+
+Proposed API group: `policy/v1alpha1`
+
+
+### SchedulingPolicy content
+
+SchedulingPolicy spec is composed of optional fields that allow scheduling rules. If a field is absent from a SchedulingPolicy, this `SchedulingPolicy` won't allow any item from the missing field.
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: SchedulingPolicy
+metadata:
+  name: mySchedulingPolicy
+spec:
+  priority: <exception,cluster,user,default>
+  namespaces:
+    matchNames:
+      - ns1
+      - ns2
+    matchLabels:
+      key: value
+  action:  <required,allowed,denied>
+  rules:
+    schedulerNames:
+       match: []
+    priorityClassNames:
+       match: []
+    tolerations:
+       match: []
+    nodeSelectors:
+       match: []
+    nodeAffinities:
+       match: []
+    podAntiAffinities:
+       match: []
+    podAffinities:
+       match: []
+```
+
+
+### required
```

Comment:
@tallclair while it may be possible to compose `required` with allow and deny rules, I think having a `required` action makes rules easier to read and manage. Without the `required` action, the rules could get complicated when a combination of multiple scheduling rules (for example, toleration, nodeSelector, nodeAffinities, ...) are required. It could get even more complicated if there is also a set of denied rules combined with the required ones.

With the `required` action, your "BestEffort" example is simplified to:

```
podSelector:
  'QoS':'BestEffort'
action: required
podAntiAffinities:
  includes: "critical"
```

Reply:
I agree with @bsalamat